### PR TITLE
Pietsmiet

### PR DESCRIPTION
**Pull request by mhartkorn**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence): [Permission by @duracell](https://github.com/mhartkorn/youtube-dl-ps/issues/2)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Adds extractor for PietSmiet.de, a German Let's Play group originally from YouTube that wants to become more independent from YouTube by creating their own website. Since a few months ago there is exclusive content available on their website that is not being uploaded to YouTube.

### CODE REVIEW

**Comment by duracell** [comment]
@rg3 Is nobody active anymore in reviewing PR?

**Comment by rg3** [comment]
Same here @dstftw

File: youtube_dl/extractor/pietsmiet.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+)
+from ..utils import (
+    unescapeHTML,
+    js_to_json,
+    int_or_none,
+)
+
+
+class PietsmietIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?pietsmiet\.de/gallery/(categories|playlists)/[\w-]+/(?P<id>\d+)-.*/?'
```
Comment (dstftw) [inline youtube_dl/extractor/pietsmiet.py]: `.*/?` is pointless at the end.

File: youtube_dl/extractor/pietsmiet.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        page_id = self._match_id(url)
+        webpage = self._download_webpage(url, page_id)
+        data_video_config = self._search_regex(
+            r'var config =(.*?)\};\n', webpage, 'video config', flags=re.DOTALL)
+        data_video_config = data_video_config.replace(']', '],', 1) + '}'
+        data_video = self._parse_json(js_to_json(unescapeHTML(data_video_config)), page_id)
+
+        formats = []
+
+        m3u8_manifest_urls = filter(lambda x: x['file'].endswith('m3u8'), data_video['sources'])
```
Comment (dstftw) [inline youtube_dl/extractor/pietsmiet.py]: Single loop for all sources without any unnecessary intermediates.

File: youtube_dl/extractor/pietsmiet.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        }
+    ]
+
+    def _real_extract(self, url):
+        page_id = self._match_id(url)
+        webpage = self._download_webpage(url, page_id)
+        data_video_config = self._search_regex(
+            r'var config =(.*?)\};\n', webpage, 'video config', flags=re.DOTALL)
+        data_video_config = data_video_config.replace(']', '],', 1) + '}'
```
Comment (dstftw) [inline youtube_dl/extractor/pietsmiet.py]: This should be fixed in `js_to_json`.

File: youtube_dl/extractor/pietsmiet.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                    '([0-9]+)p', label, 'http video height',
+                    default=720, fatal=False)
+                format_height = int_or_none(format_height_raw)
+
+            if format_height > 0:
+                format_width = float(format_height) * (16 / 9)
+
+                formats.append({
+                    'format_id': 'http-{0}'.format(label),
+                    'url': "https:{0}".format(m['file']),
```
Comment (dstftw) [inline youtube_dl/extractor/pietsmiet.py]: Single quotes.

File: youtube_dl/extractor/pietsmiet.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        mp4_urls = filter(lambda x: not x['file'].endswith('m3u8'), data_video['sources'])
+        for m in mp4_urls:
+            label = m.get('label')
+            format_height = 0
+
+            if label:
+                # Calculate resolution for HTTP format but should always be 1280x720
+                format_height_raw = self._search_regex(
+                    '([0-9]+)p', label, 'http video height',
+                    default=720, fatal=False)
```
Comment (dstftw) [inline youtube_dl/extractor/pietsmiet.py]: `default` implies non fatal.

File: youtube_dl/extractor/pietsmiet.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                    'ext': m.get('type'),
+                    'fps': 30.0,
+                })
+
+        self._sort_formats(formats)
+
+        return {
+            'id': page_id,
+            'display_id': page_id,
+            'title': compat_urllib_parse_unquote(data_video['abouttext']),
```
Comment (dstftw) [inline youtube_dl/extractor/pietsmiet.py]: This should be extracted very first.

File: youtube_dl/extractor/pietsmiet.py (hunk 3 above)
Comment (mhartkorn) [reply youtube_dl/extractor/pietsmiet.py]: I looked into this but I'm unsure on how to do that. I don't want to fix code of this importance to the project without guidance from a project maintainer.

The current error is `ERROR: 41807: Failed to parse JSON  (caused by JSONDecodeError("Expecting ',' delimiter: line 7 column 6 (char 377)",));`

And the corresponding JSON is here (weird formatting left in tact on purpose): https://gist.github.com/mhartkorn/f081c5a38692a19444b46a2c51d3a4bb

**Pull request closed without merging**